src/image/filterRenderer2D.js

```diff
@@ -0,0 +1,189 @@
+        if (!this._emptyTexture) {
+          const im = new Image(1, 1);
+          im.set(0, 0, 255);
+          this._emptyTexture = new Texture(this._renderer, im);
+        }
+        return this._emptyTexture;
+      },
+    };
+
+    // Fragment shaders mapped to filter operations
+    this.filterShaders = {
+      [constants.BLUR]: filterBlurFrag,
+      [constants.INVERT]: filterInvertFrag,
+      [constants.THRESHOLD]: filterThresholdFrag,
+      [constants.ERODE]: filterErodeFrag,
+      [constants.GRAY]: filterGrayFrag,
+      [constants.DILATE]: filterDilateFrag,
+      [constants.POSTERIZE]: filterPosterizeFrag,
+      [constants.OPAQUE]: filterOpaqueFrag,
+    };
+
+    this._shader = null;
+    this._initializeShader();
+  }
+  
+  /**
+   * Initializes the shader program if it hasn't been already.
+   */
+  _initializeShader() {
+    if (this._shader) return; // Already initialized
+
+    const fragShaderSrc = this.filterShaders[this.operation];
+    if (!fragShaderSrc) {
+      console.error("No shader available for this operation:", this.operation);
+      return;
+    }
+
+    this._shader = new Shader(this._renderer, filterShaderVert, fragShaderSrc);
+  }
+
+  /**
+   * Binds a buffer to the drawing context
+   * when passed more than two arguments it also updates or initializes
+   * the data associated with the buffer
+   */
+  _bindBufferData(buffer, target, values, type, usage) {
+    const gl = this.gl;
+    gl.bindBuffer(target, buffer);
+    let data = values instanceof (type || Float32Array) ? values : new (type || Float32Array)(values);
+    gl.bufferData(target, data, usage || gl.STATIC_DRAW);
+  }
+
+  /**
+   * Prepares and runs the full-screen quad draw call.
+   */
+  _renderPass() {
+    const gl = this.gl;
+    this._shader.bindShader();
+
+    const pixelDensity = this.pInst._renderer.pixelDensity ? this.pInst._renderer.pixelDensity() : 1;
+
+    const texelSize = [
+      1 / (this.pInst.width * pixelDensity),
+      1 / (this.pInst.height * pixelDensity)
+    ];
+
+    const canvasTexture = new Texture(this._renderer, this.pInst._renderer.wrappedElt);
+
+    // Set uniforms for the shader
+    this._shader.setUniform('tex0', canvasTexture);
+    this._shader.setUniform('texelSize', texelSize);
+    this._shader.setUniform('canvasSize', [this.pInst.width, this.pInst.height]);
+    this._shader.setUniform('radius', Math.max(1, this.filterParameter));
+
+    // Identity matrices for projection/model-view (unsure)
+
+    // TODO: FIX IT
+    const identityMatrix = [1,0,0,0, 0,1,0,0, 0,0,1,0, 0,0,0,1];
+    this._shader.setUniform('uModelViewMatrix', identityMatrix);
+    this._shader.setUniform('uProjectionMatrix', identityMatrix);
+
+    // Set up the vertices and texture coordinates for a full-screen quad
+    const vertices = [-1, -1, 1, -1, -1, 1, 1, 1];
+    const texcoords = [0, 1, 1, 1, 0, 0, 1, 0];
+
+    // Create and bind buffers
+    const vertexBuffer = gl.createBuffer();
```

Comment:
This seems like something we don't have to recreate every time, maybe we can cache and save these?

Reply:
Sure!